src/main/scala/com/fulcrumgenomics/umi/SimpleConsensusCaller.scala

```diff
@@ -0,0 +1,76 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.umi
+
+import com.fulcrumgenomics.FgBioDef.forloop
+import com.fulcrumgenomics.util.NumericTypes.LogProbability
+
+/**
+  * A trait that can be mixed in to call sequences represented as strings of the same length.
+  *
+  * If a non-ACGTN bases is encountered at any position in the strings, that position is not consensus called, and the
+  * character from the first sequenced is used.
+  *
+  * @param errorRatePreLabeling The error probability prior to adding the UMI.  By default, set to almost surely zero.
+  * @param errorRatePostLabeling The error probability post to adding the UMI.  By default, set to almost surely zero
+  * @param pError The adjusted error probability.  By default set to ln(0.99), as the actual values don't matter much.
+  * @param pTruth The adjusted truth probability.  By default set to ln(0.01), as the actual values don't matter much.
+  */
+private[umi] class SimpleConsensusCaller(val errorRatePreLabeling: Byte = 90.toByte,
+                                         val errorRatePostLabeling: Byte = 90.toByte,
+                                         val pError: LogProbability = -4.60517,
+                                         val pTruth: LogProbability = -0.01005034
+                                        ) {
+
+  /** A consensus caller used to generate consensus UMI sequences */
+  private val consensusBuilder = new ConsensusCaller(
+    errorRatePreLabeling  = this.errorRatePreLabeling,
+    errorRatePostLabeling = this.errorRatePostLabeling
+  ).builder()
+
+  private val DnaBases = Set('A', 'C', 'G', 'T', 'N', 'a', 'c', 'g', 't', 'n')
+
+  /** Calls a simple consensus sequences from a set of sequences all the same length. */
+  def callConsensus(sequences: Seq[String]): String = {
+    require(sequences.nonEmpty, "Can't call consensus on an empty set of UMIs!")
+    require(sequences.forall(_.length == sequences.head.length), "Sequences must all have the same length")
+    val buffer = new StringBuilder
+
+    forloop(from=0, until=sequences.head.length) { i =>
+      this.consensusBuilder.reset()
+      val nonDnaBase = sequences.map(sequence => sequence.charAt(i)).flatMap { char =>
```

Comment:
Given that `sequences` is a `Seq` this is going to create two more concrete `Seq`s along the way - one for the chars, and then another for the option(s).  How about something like this instead:

```scala
forloop (from=0, until=sequences.head.length) { i =>
  this.consensusBuilder.reset()
  var nonDna = 0
  for (sequence <- sequences; char <- sequence) {
    if (!this.DnaBases.contains(char)) nonDna += 1
    else this.consensusBuilder.add(char.toByte, pError=this.pError, pTruth=this.pTruth)
  }

  if (nonDna == 0) buffer.append(this.consensusBuilder.call())
  else if (nonDna == sequences.length) buffer.append(sequences.head.charAt(i))
  else throw new IllegalStateException(s"Sequences contained a mix of DNA and non-DNA characters at offset $i: $sequences")
}
```

Reply:
Close, the inner loop is wrong (we want the base at index `i` for each string), but I get the idea.